src/sentry/analytics/event.py

```diff
@@ -1,99 +1,50 @@
 from base64 import b64encode
 from collections.abc import Mapping
+from typing import Any, Sequence
 from uuid import uuid1
 
 from django.utils import timezone
 
+from sentry.analytics.attribute import Attribute
+from sentry.analytics.utils import get_data
 from sentry.utils.dates import to_timestamp
 
 
-class Attribute:
-    def __init__(self, name, type=str, required=True):
-        self.name = name
-        self.type = type
-        self.required = required
-
-    def extract(self, value):
-        if value is None:
-            return value
-        return self.type(value)
-
-
-class Map(Attribute):
-    def __init__(self, name, attributes=None, required=True):
-        self.name = name
-        self.required = required
-        self.attributes = attributes
-
-    def extract(self, value):
-        """
-        If passed a non dictionary we assume we can pull attributes from it.
-
-        This will hard error in some situations if you're passing a bad type
-        (like an int).
-        """
-        if value is None:
-            return value
-
-        if not isinstance(value, Mapping):
-            new_value = {}
-            for attr in self.attributes:
-                new_value[attr.name] = attr.extract(getattr(value, attr.name, None))
-            items = new_value
-        else:
-            # ensure we don't mutate the original
-            # we don't need to deepcopy as if it recurses into another Map it
-            # will once again copy itself
-            items = value.copy()
-
-        data = {}
-        for attr in self.attributes:
-            nv = items.pop(attr.name, None)
-            if attr.required and nv is None:
-                raise ValueError(f"{attr.name} is required (cannot be None)")
-
-            data[attr.name] = attr.extract(nv)
-
-        if items:
-            raise ValueError("Unknown attributes: {}".format(", ".join(map(str, items.keys()))))
-
-        return data
-
-
-class Event:
+class Event(abc.ABC):
     __slots__ = ["uuid", "data", "datetime"]
 
+    # This MUST be overridden by child classes.
     type = None
 
-    attributes = ()
+    # These should be overridden by child classes.
+    attributes: Sequence[Attribute] = ()
 
-    def __init__(self, type=None, datetime=None, **items):
+    def __init__(
+        self, type: Any | None = None, datetime: dt.datetime | None = None, **items: Any
+    ) -> None:
         self.uuid = uuid1()
-
         self.datetime = datetime or timezone.now()
-        if type is not None:
-            self.type = type
+        self.type = self._get_type(type)
+        self.data = get_data(self.attributes, items)
+
+    def _get_type(self, _type: Any | None = None) -> Any:
+        """
+        The Event's `type` can either be passed in as a parameter or set as a
+        property on a child class.
+        """
+        if _type is not None:
+            return _type
 
         if self.type is None:
             raise ValueError("Event is missing type")
```

Comment:
Is `NotImplementedError` more accurate here? I know you didn't make this change but just a thought looking at your comment on L20.

Reply:
I treat `raise NotImplementedError` the same as [other languages labeling an interface's method as abstract](https://en.wikipedia.org/wiki/Abstract_type). This function is reading from parameters and doing polymorphism and is just generally hacky so I'm ok with it using a different exception. Also I didn't want to change the behavior for this refactor.

Reply:
Yeah, it makes sense not to introduce new behavior in PRs like this, good call